File: frontend/src/stores/pluginStore.ts
import { create } from "zustand";
import {
  ListPlugins,
  GetPlugin,
  SavePlugin,
  DeletePlugin,
  TogglePlugin,
  TestPlugin,
} from "../../wailsjs/go/main/App";
import { main } from "../../wailsjs/go/models";

export interface Plugin {
  metadata: {
    id: string;
    name: string;
    version: string;
    author: string;
    description: string;
    enabled: boolean;
    filters: {
      sources: string[];
      types: string[];
      levels: string[];
      urlPattern: string;
      titleMatch: string;
    };
    config: Record<string, any>;
    createdAt: string;
    updatedAt: string;
  };
  sourceCode: string;
  language: "javascript" | "typescript";
  compiledCode: string;
}

export interface TestResult {
  derivedEvents: any[];
  tags: string[];
  metadata: Record<string, any>;
  error?: string;
}

interface PluginState {
  // 插件列表
  plugins: Plugin[];
  loading: boolean;
  error: string | null;

  // 当前编辑的插件
  currentPlugin: Plugin | null;
  editorOpen: boolean;

  // 测试相关
  testResult: TestResult | null;
  testing: boolean;

  // 操作
  loadPlugins: () => Promise<void>;
  getPlugin: (id: string) => Promise<Plugin | null>;
  savePlugin: (plugin: Partial<Plugin>) => Promise<void>;
  deletePlugin: (id: string) => Promise<void>;
  togglePlugin: (id: string, enabled: boolean) => Promise<void>;
  testPlugin: (script: string, eventId: string) => Promise<TestResult>;

  // UI 操作
  openEditor: (plugin?: Plugin) => void;
  closeEditor: () => void;
  setCurrentPlugin: (plugin: Plugin | null) => void;
}

export const usePluginStore = create<PluginState>((set, get) => ({
  // 状态
  plugins: [],
  loading: false,
  error: null,
  currentPlugin: null,
  editorOpen: false,
  testResult: null,
  testing: false,

  // 加载插件列表
  loadPlugins: async () => {
    set({ loading: true, error: null });
    try {
      const result = await ListPlugins();
      // ListPlugins 返回 PluginMetadata[]，需要转换为 Plugin[] 格式
      const plugins: Plugin[] = (result as any[]).map((metadata: any) => ({
        metadata: {
          id: metadata.id || "",
          name: metadata.name || "",
          version: metadata.version || "1.0.0",
          author: metadata.author || "",
          description: metadata.description || "",
          enabled: metadata.enabled !== undefined ? metadata.enabled : true,
          filters: metadata.filters || { sources: [], types: [], levels: [], urlPattern: "", titleMatch: "" },
          config: metadata.config || {},
          createdAt: metadata.createdAt || "",
          updatedAt: metadata.updatedAt || "",
        },
        sourceCode: "", // ListPlugins 不返回源码，需要时调用 GetPlugin
        language: "typescript" as const,
        compiledCode: "",
      }));
      set({ plugins, loading: false });
    } catch (error) {
      console.error("Failed to load plugins:", error);
      set({
        error: error instanceof Error ? error.message : "Failed to load plugins",
        loading: false,
      });
    }
  },

  // 获取单个插件
  getPlugin: async (id: string) => {
    set({ loading: true, error: null });
    try {
      const result = await GetPlugin(id);
      const plugin = result as any;
      // GetPlugin 返回完整的 Plugin 对象（已包含 metadata 嵌套结构）
      const pluginData: Plugin = {
        metadata: plugin.metadata || {},
        sourceCode: plugin.sourceCode || "",
        language: (plugin.language || "typescript") as "javascript" | "typescript",
        compiledCode: plugin.compiledCode || "",
      };
      set({ loading: false });
      return pluginData;
    } catch (error) {
      console.error("Failed to get plugin:", error);
      set({
        error: error instanceof Error ? error.message : "Failed to get plugin",
        loading: false,
      });
      return null;
    }
  },

  // 保存插件
  savePlugin: async (plugin: Partial<Plugin>) => {
    set({ loading: true, error: null });
    try {
      // 使用 Wails 生成的 PluginSaveRequest 类
      const req = new main.PluginSaveRequest({
        id: plugin.metadata?.id || "",
        name: plugin.metadata?.name || "",
        version: plugin.metadata?.version || "1.0.0",
        author: plugin.metadata?.author || "",
        description: plugin.metadata?.description || "",
        sourceCode: plugin.sourceCode || "",
        language: plugin.language || "typescript",
        compiledCode: plugin.compiledCode || plugin.sourceCode || "",
        filters: plugin.metadata?.filters || { sources: [], types: [], levels: [], urlPattern: "", titleMatch: "" },
        config: plugin.metadata?.config || {},
      });

      await SavePlugin(req);
      await get().loadPlugins();
      set({ loading: false, editorOpen: false, currentPlugin: null });
    } catch (error) {
      console.error("Failed to save plugin:", error);
      set({
        error: error instanceof Error ? error.message : "Failed to save plugin",
        loading: false,
      });
      throw error;
    }
  },

  // 删除插件
  deletePlugin: async (id: string) => {
    set({ loading: true, error: null });
    try {
      await DeletePlugin(id);
      await get().loadPlugins();
      set({ loading: false });
    } catch (error) {
      console.error("Failed to delete plugin:", error);
      set({
        error: error instanceof Error ? error.message : "Failed to delete plugin",
        loading: false,
      });
      throw error;
    }
  },

  // 启用/禁用插件
  togglePlugin: async (id: string, enabled: boolean) => {
    try {
      await TogglePlugin(id, enabled);
      // 更新本地状态
      set((state) => ({
        plugins: state.plugins.map((p) =>
          p.metadata.id === id
            ? { ...p, metadata: { ...p.metadata, enabled } }
            : p
        ),
      }));
    } catch (error) {
      console.error("Failed to toggle plugin:", error);
      set({
        error: error instanceof Error ? error.message : "Failed to toggle plugin",
      });
      throw error;
    }
  },

  // 测试插件
  testPlugin: async (script: string, eventId: string) => {
    set({ testing: true, testResult: null, error: null });
    try {
      const result = await TestPlugin(script, eventId);
      const testResult = result as any;
      set({ testResult, testing: false });
      return testResult;
    } catch (error) {
      console.error("Failed to test plugin:", error);
      set({
        error: error instanceof Error ? error.message : "Failed to test plugin",
        testing: false,
      });
      throw error;
    }
  },

  // UI 操作
  openEditor: (plugin?: Plugin) => {
    set({
      editorOpen: true,
      currentPlugin: plugin || null,
      error: null,
    });
  },

  closeEditor: () => {
    set({
      editorOpen: false,
      currentPlugin: null,
      error: null,
    });
  },

  setCurrentPlugin: (plugin: Plugin | null) => {
    set({ currentPlugin: plugin });
  },
}));

File: plugin_store.go
package main

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"log"
	"time"
)

// PluginStore 插件数据库存储
type PluginStore struct {
	db *sql.DB
}

// NewPluginStore 创建插件存储
func NewPluginStore(db *sql.DB) *PluginStore {
	return &PluginStore{db: db}
}

// InitSchema 初始化数据库表
func (ps *PluginStore) InitSchema() error {
	schema := `
	CREATE TABLE IF NOT EXISTS plugins (
		id TEXT PRIMARY KEY,
		name TEXT NOT NULL,
		version TEXT NOT NULL,
		author TEXT,
		description TEXT,
		
		-- 源代码
		source_code TEXT NOT NULL,
		source_language TEXT DEFAULT 'typescript',
		compiled_code TEXT NOT NULL,
		
		-- 配置
		filters TEXT,
		config TEXT,
		
		-- 状态
		enabled INTEGER DEFAULT 1,
		
		-- 时间戳
		created_at INTEGER NOT NULL,
		updated_at INTEGER NOT NULL
	);

	CREATE TABLE IF NOT EXISTS plugin_versions (
		id INTEGER PRIMARY KEY AUTOINCREMENT,
		plugin_id TEXT NOT NULL,
		version TEXT NOT NULL,
		name TEXT NOT NULL,
		author TEXT,
		description TEXT,
		source_code TEXT NOT NULL,
		source_language TEXT DEFAULT 'typescript',
		compiled_code TEXT NOT NULL,
		filters TEXT,
		config TEXT,
		saved_at INTEGER NOT NULL,
		FOREIGN KEY (plugin_id) REFERENCES plugins(id) ON DELETE CASCADE
	);

	CREATE INDEX IF NOT EXISTS idx_plugin_versions_plugin_id ON plugin_versions(plugin_id);
	CREATE INDEX IF NOT EXISTS idx_plugin_versions_saved_at ON plugin_versions(saved_at);

	-- plugins 表索引
	CREATE INDEX IF NOT EXISTS idx_plugins_enabled ON plugins(enabled);
	CREATE INDEX IF NOT EXISTS idx_plugins_created_at ON plugins(created_at);
	CREATE INDEX IF NOT EXISTS idx_plugins_updated_at ON plugins(updated_at);
	`

	_, err := ps.db.Exec(schema)
	return err
}

// SavePlugin 保存插件
func (ps *PluginStore) SavePlugin(plugin *Plugin) error {
	filtersJSON, err := json.Marshal(plugin.Metadata.Filters)
	if err != nil {
		return fmt.Errorf("序列化 filters 失败: %w", err)
	}

	configJSON, err := json.Marshal(plugin.Metadata.Config)
	if err != nil {
		return fmt.Errorf("序列化 config 失败: %w", err)
	}

	now := time.Now().Unix()
	enabled := 0
	if plugin.Metadata.Enabled {
		enabled = 1
	}

	// 检查是否已存在
	var exists int
	err = ps.db.QueryRow("SELECT COUNT(*) FROM plugins WHERE id = ?", plugin.Metadata.ID).Scan(&exists)
	if err != nil {
		return err
	}

	if exists > 0 {
		// 使用事务保护更新操作（版本历史 + 更新当前记录）
		tx, txErr := ps.db.Begin()
		if txErr != nil {
			return fmt.Errorf("开始事务失败: %w", txErr)
		}
		defer func() {
			if txErr != nil {
				tx.Rollback()
			}
		}()

		// 更新前先保存旧版本到历史表
		var oldPlugin Plugin
		var oldFiltersJSON, oldConfigJSON string
		var oldEnabled int
		err = tx.QueryRow(`
			SELECT name, version, author, description,
				   source_code, source_language, compiled_code,
				   filters, config, enabled
			FROM plugins WHERE id = ?
		`, plugin.Metadata.ID).Scan(
			&oldPlugin.Metadata.Name,
			&oldPlugin.Metadata.Version,
			&oldPlugin.Metadata.Author,
			&oldPlugin.Metadata.Description,
			&oldPlugin.SourceCode,
			&oldPlugin.Language,
			&oldPlugin.CompiledCode,
			&oldFiltersJSON,
			&oldConfigJSON,
			&oldEnabled,
		)

		if err == nil {
			// 保存旧版本到历史表
			_, err = tx.Exec(`
				INSERT INTO plugin_versions (
					plugin_id, version, name, author, description,
					source_code, source_language, compiled_code,
					filters, config, saved_at
				) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
			`,
				plugin.Metadata.ID,
				oldPlugin.Metadata.Version,
				oldPlugin.Metadata.Name,
				oldPlugin.Metadata.Author,
				oldPlugin.Metadata.Description,
				oldPlugin.SourceCode,
				oldPlugin.Language,
				oldPlugin.CompiledCode,
				oldFiltersJSON,
				oldConfigJSON,
				now,
			)
			if err != nil {
				log.Printf("[PluginStore] Failed to save old version: %v", err)
				// 继续执行更新，不因历史保存失败而阻止更新
			}

			// 自动清理旧版本：保留最近 maxVersions 个
			ps.pruneVersions(plugin.Metadata.ID, 20)
		}

		// 更新
		_, txErr = tx.Exec(`
			UPDATE plugins SET
				name = ?,
				version = ?,
				author = ?,
				description = ?,
				source_code = ?,
				source_language = ?,
				compiled_code = ?,
				filters = ?,
				config = ?,
				enabled = ?,
				updated_at = ?
			WHERE id = ?
		`,
			plugin.Metadata.Name,
			plugin.Metadata.Version,
			plugin.Metadata.Author,
			plugin.Metadata.Description,
			plugin.SourceCode,
			plugin.Language,
			plugin.CompiledCode,
			string(filtersJSON),
			string(configJSON),
			enabled,
			now,
			plugin.Metadata.ID,
		)
		if txErr != nil {
			return txErr
		}

		return tx.Commit()
	} else {
		// 插入
		_, err = ps.db.Exec(`
			INSERT INTO plugins (
				id, name, version, author, description,
				source_code, source_language, compiled_code,
				filters, config, enabled,
				created_at, updated_at
			) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
		`,
			plugin.Metadata.ID,
			plugin.Metadata.Name,
			plugin.Metadata.Version,
			plugin.Metadata.Author,
			plugin.Metadata.Description,
			plugin.SourceCode,
			plugin.Language,
			plugin.CompiledCode,
			string(filtersJSON),
			string(configJSON),
			enabled,
			now,
			now,
		)
	}

	return err
}

// GetPlugin 获取插件
func (ps *PluginStore) GetPlugin(id string) (*Plugin, error) {
	var (
		name         string
		version      string
		author       string
		description  string
		sourceCode   string
		language     string
		compiledCode string
		filtersJSON  string
		configJSON   string
		enabled      int
		createdAt    int64
		updatedAt    int64
	)

	err := ps.db.QueryRow(`
		SELECT name, version, author, description,
			   source_code, source_language, compiled_code,
			   filters, config, enabled,
			   created_at, updated_at
		FROM plugins
		WHERE id = ?
	`, id).Scan(
		&name, &version, &author, &description,
		&sourceCode, &language, &compiledCode,
		&filtersJSON, &configJSON, &enabled,
		&createdAt, &updatedAt,
	)

	if err == sql.ErrNoRows {
		return nil, fmt.Errorf("插件不存在: %s", id)
	}
	if err != nil {
		return nil, err
	}

	var filters PluginFilters
	if err := json.Unmarshal([]byte(filtersJSON), &filters); err != nil {
		return nil, fmt.Errorf("解析 filters 失败: %w", err)
	}

	var config map[string]interface{}
	if err := json.Unmarshal([]byte(configJSON), &config); err != nil {
		return nil, fmt.Errorf("解析 config 失败: %w", err)
	}

	plugin := &Plugin{
		Metadata: PluginMetadata{
			ID:          id,
			Name:        name,
			Version:     version,
			Author:      author,
			Description: description,
			Enabled:     enabled == 1,
			Filters:     filters,
			Config:      config,
			CreatedAt:   time.Unix(createdAt, 0),
			UpdatedAt:   time.Unix(updatedAt, 0),
		},
		SourceCode:   sourceCode,
		Language:     language,
		CompiledCode: compiledCode,
		State:        make(map[string]interface{}),
	}

	return plugin, nil
}

// ListPlugins 列出所有插件
func (ps *PluginStore) ListPlugins() ([]*Plugin, error) {
	rows, err := ps.db.Query(`
		SELECT id, name, version, author, description,
			   source_code, source_language, compiled_code,
			   filters, config, enabled,
			   created_at, updated_at
		FROM plugins
		ORDER BY created_at DESC
	`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var plugins []*Plugin

	for rows.Next() {
		var (
			id           string
			name         string
			version      string
			author       string
			description  string
			sourceCode   string
			language     string
			compiledCode string
			filtersJSON  string
			configJSON   string
			enabled      int
			createdAt    int64
			updatedAt    int64
		)

		err := rows.Scan(
			&id, &name, &version, &author, &description,
			&sourceCode, &language, &compiledCode,
			&filtersJSON, &configJSON, &enabled,
			&createdAt, &updatedAt,
		)
		if err != nil {
			return nil, err
		}

		var filters PluginFilters
		if err := json.Unmarshal([]byte(filtersJSON), &filters); err != nil {
			log.Printf("[PluginStore] Skipping plugin %s: invalid filters JSON: %v", id, err)
			continue
		}

		var config map[string]interface{}
		if err := json.Unmarshal([]byte(configJSON), &config); err != nil {
			log.Printf("[PluginStore] Skipping plugin %s: invalid config JSON: %v", id, err)
			continue
		}

		plugin := &Plugin{
			Metadata: PluginMetadata{
				ID:          id,
				Name:        name,
				Version:     version,
				Author:      author,
				Description: description,
				Enabled:     enabled == 1,
				Filters:     filters,
				Config:      config,
				CreatedAt:   time.Unix(createdAt, 0),
				UpdatedAt:   time.Unix(updatedAt, 0),
			},
			SourceCode:   sourceCode,
			Language:     language,
			CompiledCode: compiledCode,
			State:        make(map[string]interface{}),
		}

		plugins = append(plugins, plugin)
	}

	return plugins, rows.Err()
}

// DeletePlugin 删除插件
func (ps *PluginStore) DeletePlugin(id string) error {
	_, err := ps.db.Exec("DELETE FROM plugins WHERE id = ?", id)
	return err
}

// pruneVersions 清理旧版本，保留最近 maxVersions 个
func (ps *PluginStore) pruneVersions(pluginID string, maxVersions int) {
	_, err := ps.db.Exec(`
		DELETE FROM plugin_versions
		WHERE plugin_id = ? AND id NOT IN (
			SELECT id FROM plugin_versions
			WHERE plugin_id = ?
			ORDER BY saved_at DESC
			LIMIT ?
		)
	`, pluginID, pluginID, maxVersions)
	if err != nil {
		log.Printf("[PluginStore] Failed to prune old versions for %s: %v", pluginID, err)
	}
}

// PluginVersion 插件历史版本
type PluginVersion struct {
	ID          int       `json:"id"`
	PluginID    string    `json:"pluginId"`
	Version     string    `json:"version"`
	Name        string    `json:"name"`
	Author      string    `json:"author"`
	Description string    `json:"description"`
	SavedAt     time.Time `json:"savedAt"`
}

// GetPluginVersions 获取插件的历史版本列表
func (ps *PluginStore) GetPluginVersions(pluginID string, limit int) ([]PluginVersion, error) {
	if limit <= 0 {
		limit = 10
	}

	rows, err := ps.db.Query(`
		SELECT id, plugin_id, version, name, author, description, saved_at
		FROM plugin_versions
		WHERE plugin_id = ?
		ORDER BY saved_at DESC
		LIMIT ?
	`, pluginID, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var versions []PluginVersion
	for rows.Next() {
		var v PluginVersion
		var savedAt int64
		err := rows.Scan(&v.ID, &v.PluginID, &v.Version, &v.Name, &v.Author, &v.Description, &savedAt)
		if err != nil {
			continue
		}
		v.SavedAt = time.Unix(savedAt, 0)
		versions = append(versions, v)
	}

	return versions, rows.Err()
}

// RollbackPlugin 回滚到指定版本
func (ps *PluginStore) RollbackPlugin(versionID int) error {
	// 获取历史版本数据
	var (
		pluginID     string
		version      string
		name         string
		author       string
		description  string
		sourceCode   string
		language     string
		compiledCode string
		filtersJSON  string
		configJSON   string
	)

	err := ps.db.QueryRow(`
		SELECT plugin_id, version, name, author, description,
			   source_code, source_language, compiled_code,
			   filters, config
		FROM plugin_versions
		WHERE id = ?
	`, versionID).Scan(
		&pluginID, &version, &name, &author, &description,
		&sourceCode, &language, &compiledCode,
		&filtersJSON, &configJSON,
	)
	if err != nil {
		return fmt.Errorf("历史版本不存在: %w", err)
	}

	// 先保存当前版本到历史（回滚前备份）
	now := time.Now().Unix()
	var currentPlugin Plugin
	var currentFiltersJSON, currentConfigJSON string
	var currentEnabled int

	err = ps.db.QueryRow(`
		SELECT name, version, author, description,
			   source_code, source_language, compiled_code,
			   filters, config, enabled
		FROM plugins WHERE id = ?
	`, pluginID).Scan(
		&currentPlugin.Metadata.Name,
		&currentPlugin.Metadata.Version,
		&currentPlugin.Metadata.Author,
		&currentPlugin.Metadata.Description,
		&currentPlugin.SourceCode,
		&currentPlugin.Language,
		&currentPlugin.CompiledCode,
		&currentFiltersJSON,
		&currentConfigJSON,
		&currentEnabled,
	)

	if err == nil {
		// 保存当前版本到历史
		_, _ = ps.db.Exec(`
			INSERT INTO plugin_versions (
				plugin_id, version, name, author, description,
				source_code, source_language, compiled_code,
				filters, config, saved_at
			) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
		`,
			pluginID,
			currentPlugin.Metadata.Version,
			currentPlugin.Metadata.Name,
			currentPlugin.Metadata.Author,
			currentPlugin.Metadata.Description,
			currentPlugin.SourceCode,
			currentPlugin.Language,
			currentPlugin.CompiledCode,
			currentFiltersJSON,
			currentConfigJSON,
			now,
		)
	}

	// 更新插件为历史版本
	_, err = ps.db.Exec(`
		UPDATE plugins SET
			name = ?,
			version = ?,
			author = ?,
			description = ?,
			source_code = ?,
			source_language = ?,
			compiled_code = ?,
			filters = ?,
			config = ?,
			updated_at = ?
		WHERE id = ?
	`,
		name, version, author, description,
		sourceCode, language, compiledCode,
		filtersJSON, configJSON,
		now, pluginID,
	)

	return err
}
